ttl: add checkpointing to TTL job

spilchen opened the pull request:
Introduce a progress tracker for TTL jobs that supports checkpointing. Previously, if a TTL job restarted, it would reprocess spans that had already been completed. With this change, completed spans are stored in the job record so that restarts can skip over them, avoiding duplicate work.

Fixes #140514
Epic: none

Release note (performance improvement): TTL jobs now checkpoint their progress, allowing them to resume without reprocessing already completed spans after a restart.

rafiss reviewed (changes requested):
the new checkpoint logic looks good! i just had one comment about the flushing goroutine and other minor suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-:-O_Xm49O6QX9cSIAI2d1:bo6krgo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)

___
*[`pkg/sql/ttl/ttljob/progress.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XhjgfBat56HzjyV78:-O_XhjgfBat56HzjyV79:b-20zq6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/sql/ttl/ttljob/progress.go#L63)):*
> ```Go
> 	handleProgressUpdate(ctx context.Context, meta *execinfrapb.ProducerMetadata) error
> 	// termTracker performs any necessary cleanup when the job completes or fails.
> 	termTracker()
> ```

super nit: `terminateTracker` might be more clear -- only saying so since I initially interpreted this as the word "term"
___
*[`pkg/sql/ttl/ttljob/progress.go` line 491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XkOmY6EQ6dgmLwrN3:-O_XkOmY6EQ6dgmLwrN4:byamirl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/sql/ttl/ttljob/progress.go#L491)):*
> ```Go
> 			case <-timer.Ch():
> 				if err := write(ctx); err != nil {
> 					return errors.Wrap(err, "could not flush progress")
> ```

is returning the error desired here? asking since we had a similar issue in the backfiller, where we returned a similar error but that meant that the first time there was any issue writing progress, the goroutine would die and we'd never attempt to write progress again after that. (see e1e733b7630)
___
*[`pkg/sql/ttl/ttljob/progress_test.go` line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XlsT-7UtO7VlHlZuc:-O_XlsT-7UtO7VlHlZud:b5qgzok) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/sql/ttl/ttljob/progress_test.go#L287)):*
> ```Go
> }
> 
> func TestCheckpointProgressUpdaterContents(t *testing.T) {
> ```

this is a nice test!
___
*[`pkg/jobs/jobspb/jobs.proto` line 1326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XhHPF9Vy_Os2pu7MB:-O_XhHPF9Vy_Os2pu7MC:bt0x9rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/jobs/jobspb/jobs.proto#L1326)):*
> ```proto
>   // UseCheckpointing is true if the TTL job will use checkpointing for progress
>   // tracking.
>   bool use_checkpointing = 6;
> ```

super nit: could this new field be added at the end of the protobuf definition?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152618)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/jobs/jobspb/jobs.proto` line 1326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XhHPF9Vy_Os2pu7MB:-O_lVER2Ao1OZ70C_LXG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/jobs/jobspb/jobs.proto#L1326)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: could this new field be added at the end of the protobuf definition?
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/progress.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XhjgfBat56HzjyV78:-O_lVKNvD70wUz2uJuqN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/sql/ttl/ttljob/progress.go#L63)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: `terminateTracker` might be more clear -- only saying so since I initially interpreted this as the word "term"
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/progress.go` line 491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-O_XkOmY6EQ6dgmLwrN3:-O_lVNLZ7LEr1C8QcNji:b-eiwvvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0366c5a5ce84c1f1e076cbf6e6ba8da201ab4ca7/pkg/sql/ttl/ttljob/progress.go#L491)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is returning the error desired here? asking since we had a similar issue in the backfiller, where we returned a similar error but that meant that the first time there was any issue writing progress, the goroutine would die and we'd never attempt to write progress again after that. (see e1e733b7630)
</blockquote></details>

Good point. I'll log the error and continue.


<!-- Sent from Reviewable.io -->

spilchen committed:
ttl: add checkpointing to TTL job

Introduce a progress tracker for TTL jobs that supports checkpointing.
Previously, if a TTL job restarted, it would reprocess spans that had
already been completed. With this change, completed spans are stored
in the job record so that restarts can skip over them, avoiding
duplicate work.

Fixes #140514
Epic: none

Release note (performance improvement): TTL jobs now checkpoint their
progress, allowing them to resume without reprocessing already
completed spans after a restart.

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152618#-:-O_wPa1w5Z0i3WrTrWCP:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kev-cao and @msbutler)



<!-- Sent from Reviewable.io -->

spilchen commented on pkg/sql/ttl/ttljob/progress.go:
```diff
@@ -186,3 +226,401 @@ func (t *legacyProgressTracker) handleProgressUpdate(
+			return err
+		}
+		defer frontier.Release()
+		return jobfrontier.Store(ctx, txn, t.jobID, ttlCompletedSpansKey, frontier)
```
I spoke with Rafi further. We won’t backport this to 25.4. Since it has been open for a while, I don’t think it needs to wait for additional API changes to jobfrontier. We can handle those in a follow-up.

spilchen replied:
I opened a separate issue: #154062
